Add $XmlLogPath parameter to allow for specifying a custom location for the temporary WinSCP XML log file.

AndrewDP23 opened the pull request:
Added $XmlLogPath parameter to allow for specifying a custom location for the XML Log Path. This is a workaround for issues I am encountering where the default Windows temp path cannot be written to.

AndrewDP23 committed:
Update New-WinSCPSession.ps1

Added $XmlLogPath parameter to allow for specifying a custom location for the XML Log Path. This is a workaround for issues I am encountering where the default Windows temp path cannot be written to.

tomohulk commented:
This is a good find, and thank you for sharing.  However, the way you implemented it added some redundant code that isn't needed.  If you would update your request, I will merge it.

1. Please add some validation checking to your `$XmlLogPath` parameter (see the `$SessionLogPath` parameter for what I mean) to ensure if a custom path is provided, its a valid path.
2. if you look at this code:

```PowerShell
                # Enumerate each parameter.
                $sessionObjectProperties = $session |
                    Get-Member -MemberType Property |
                        Select-Object -ExpandProperty Name
                $keys = ($PSBoundParameters.Keys).Where({
                    $_ -in $sessionObjectProperties
                })

                foreach ($key in $keys) {
                    $session.$key = $PSBoundParameters.$key
                }
```

basically the code loops through every property on the `WinSCP.Session` object and maps them to the corresponding Function Parameter that is passed in the function call.  Assuming the parameter name is the exact same as the property name on the object.  So as long as the property name on a `WinSCP.Session` is `XmlLogPath`, then the code you added to map the property can be removed, and that block of code would add it automatically.

Thanks again for your request.

tomohulk commented:
Also, diving into the `WinSCP.Session` object more, should add the `$XmlLogPreserve` Argument as well

<img width="653" alt="image" src="https://github.com/tomohulk/WinSCP/assets/52005101/13f53911-689c-452f-acf7-3aae905e11fe">

tomohulk commented:
@martinprikyl why aren't these XML log properties listed on the docs page for the `WinSCP.Session` object properties?

AndrewDP23 committed:
Update New-WinSCPSession.ps1

Removed redundant code setting the session XmlLogPath property in response to pull request feedback from tomohulk. The lines on 93-98 set all properties on $session as long as they have the same name as the function parameter name.

AndrewDP23 committed:
Add $XmlLogPreserve parameter to New-WinSCPSession to allow setting of this session property in the function call.

Added $XmlLogPreserve to allow for specifying whether we want the XML log file to be retained or removed after process completion.

AndrewDP23 committed:
Added valid path validation to the $XmlLogPath parameter.

Added validation to the new $XmlLogPath parameter using the method used to validate that it is a valid file path. Validation script is a copy of an above validation script.

AndrewDP23 commented:
Thanks for this @tomohulk, I've added the path validation to the $XmlLogPath parameter and have added the additional $XmlLogPreserve parameter for review. 

Apologies for having three check-ins for this. Please let me know if you need me to cancel this and commit both the $XmlLogPath and validation script in a single check-in.

martinprikryl commented:
> @martinprikyl why aren't these XML log properties listed on the docs page for the `WinSCP.Session` object properties?

Not sure :) I'll document them.
